superduperdb/components/model.py

```diff
@@ -90,361 +85,364 @@ def get(self, k, default=None):
-                assert isinstance(X, dict)
-                X = X[key]
-            else:
-                X = [r[key] for r in X]
-        elif isinstance(key, list):
-            X = (
-                ((X[k] for k in key) if one else [(r[k] for k in key) for r in X])
-                if key
-                else X,
-            )
-        elif isinstance(key, dict):
-            X = (
-                (
-                    (X[k] for k in key.values())
-                    if one
-                    else [(r[k] for k in key.values()) for r in X]
-                )
-                if key
-                else X,
-            )
+class Mapping:
+    def __init__(self, mapping: ModelInputType, signature: str):
+        self.mapping = self._map_args_kwargs(mapping)
+        self.signature = signature
 
+    @property
+    def id_key(self):
+        out = []
+        for arg in self.mapping[0]:
+            out.append(arg)
+        for k, v in self.mapping[1]:
+            if k.startswith('_outputs.'):
+                k = k.split('.')[1]
+            out.append(f'{k}={v}')
+        return ','.join(out)
+
+    @staticmethod
+    def _map_args_kwargs(mapping):
+        if isinstance(mapping, str):
+            return ([mapping], {})
+        elif isinstance(mapping, (list, tuple)) and isinstance(mapping[0], str):
+            return (mapping, {})
+        elif isinstance(mapping, dict):
+            return ((), mapping)
         else:
-            if key is not None:
-                raise TypeError
-        return X
+            assert isinstance(mapping[0], (list, tuple))
+            assert isinstance(mapping[1], dict)
+            return mapping
 
-    def predict(
-        self,
-        X: XType,
-        db: t.Optional[Datalayer] = None,
-        select: t.Optional[CompoundSelect] = None,
-        ids: t.Optional[t.List[str]] = None,
-        max_chunk_size: t.Optional[int] = None,
-        dependencies: t.Sequence[Job] = (),
-        listen: bool = False,
-        one: bool = False,
-        context: t.Optional[t.Dict] = None,
-        insert_to: t.Optional[t.Union[TableOrCollection, str]] = None,
-        key: t.Optional[t.Union[t.Dict, t.List, str]] = None,
-        in_memory: bool = True,
-        overwrite: bool = False,
-        **kwargs,
-    ) -> t.Any:
-        was_added = self.db is not None
-        db = self.db or db
+    def __call__(self, r):
+        """
+        >>> r = {'a': 1, 'b': 2}
+        >>> self.mapping = [('a', 'b'), {}]
+        >>> _Predictor._data_from_input_type(docs)
+        ([1, 2], {})
+        >>> self.mapping = [('a',), {'b': 'X'}]
+        >>> _Predictor._data_from_input_type(docs)
+        ([1], {'X': 2})
+        """
+        args = []
+        kwargs = {}
+        for key in self.mapping[0]:
+            args.append(r[key])
+        for k, v in self.mapping[1].items():
+            kwargs[v] = r[k]
+        args = Document({'_base': args}).unpack()
+        kwargs = Document(kwargs).unpack()
```

Comment:
nice so non native datatype are supported in signature ?

Reply:
Yes.